erpnext/tests/test_couponcode.py

```diff
@@ -0,0 +1,156 @@
+        pricing_rule.warehouse=warehouse
+        pricing_rule.coupon_code_based=1
+        pricing_rule.selling=1
+        pricing_rule.append("items", {
+            "item_code": "Tesla Car",
+        })
+        pricing_rule.rate_or_discount='Discount Percentage'
+        pricing_rule.discount_percentage=20
+        pricing_rule.for_price_list='Standard Selling'
+        pricing_rule.save()
+        frappe.db.commit()
+
+@frappe.whitelist()
+def create_coupon_code(coupon_code_name=None):
+    coupon_code_name = coupon_code_name or "SAVE20"
+    if not frappe.db.exists('Coupon Code', coupon_code_name):
+        coupon_code = frappe.new_doc("Coupon Code")
+        coupon_code.coupon_name=coupon_code_name
+        coupon_code.price_list='Standard Selling'
+        coupon_code.pricing_rule="Pricing Tesla Car"
+        coupon_code.valid_from=today()
+        # getdate(add_days(today(), 3000))
+        coupon_code.maximum_use=50
+        coupon_code.save()
+        frappe.db.commit()
+
+@frappe.whitelist()
+def create_sales_partner(sales_partner_name=None):
+    sales_partner_name = sales_partner_name or "SPEEDO SALE"
+    if not frappe.db.exists('Sales Partner', sales_partner_name):
+        sales_partner = frappe.new_doc("Sales Partner")
+        sales_partner.partner_name=sales_partner_name
+        sales_partner.commission_rate=2
+        sales_partner.referral_code="SPEEDO"
+        sales_partner.save()
+        frappe.db.commit()
+
+@frappe.whitelist()
+def create_braintree_settings(braintree_settings_name=None):
+    braintree_settings_name = braintree_settings_name or "sandbox_braintree"
+    if not frappe.db.exists('Braintree Settings', braintree_settings_name):
+        braintree_settings = frappe.new_doc("Braintree Settings")
+        braintree_settings.gateway_name=braintree_settings_name
+        braintree_settings.merchant_id="test123"
+        braintree_settings.public_key="anything123"
+        braintree_settings.private_key='va_8J2ny4Naokqbpe'
+        braintree_settings.use_sandbox=1
+        braintree_settings.save()
+        frappe.db.commit()
+
+@frappe.whitelist()
+def update_shopping_cart_settings():
+    cart=frappe.get_doc('Shopping Cart Settings',None)
+    company=frappe.db.get_single_value('Global Defaults', 'default_company')
+    cart.enabled=1
+    cart.company=company
+    cart.default_customer_group='All Customer Groups'
+    cart.quotation_series=frappe.get_meta("Quotation").get_options("naming_series")
+    cart.show_price=1
+    cart.show_apply_coupon_code_in_website=1
+    cart.price_list='Standard Selling'
+    cart.enable_checkout=1
+    payment_gateway_account=frappe.get_all('Payment Gateway Account',filters={'name':("like", 'Braintree-sandbox_braintree%') }, limit=1)[0].name
+    cart.payment_gateway_account=payment_gateway_account
+    cart.save()
+    frappe.db.commit()
+
+@frappe.whitelist()
+def create_web_user(web_user_name=None):
```

Comment:
server tests should be done server-side. Such whitelisted functions are a security issue

Reply:
Hi Rushabh
-- function create_web_user(test_couponcode.py) is part of python seed/test data that is run before cypress use case is executed.
-- I was trying to "Write tests that don’t depend on each other and can be tested in isolation." [cypress_discuss_link](https://discuss.erpnext.com/t/bringing-back-ui-tests-using-cypress/43232)
-- What is the suggested way for creating test user/seed date. Please guide. thanks

Reply:
I have removed all not required whitelisting of functions. Please review. thanks!  @rmehta